instant.ts

```diff
@@ -98,47 +99,88 @@ async function runTests(path: string) {
   }
 }
 
-const orderPackageIds = (allPackages, chosenPackageIds) => {
-  function resolveDeps(id, currentStack) {
-    if (currentStack.includes(id))
-      throw Error(`Circular dependency present for id ${id}`)
-    currentStack.push(id)
+export const createDependencyTree = (allPackages, chosenPackageIds) => {
+  const tree = {}
+  const visited = new Set()
 
-    if (allPackages[id] && allPackages[id].metadata) {
-      if (
-        !allPackages[id].metadata.dependencies ||
-        !allPackages[id].metadata.dependencies.length
+  const addDependencies = (id, node) => {
+    if (visited.has(id)) {
+      throw new Error(
+        `Circular dependency detected: ${id} has already been visited.`
       )
-        return [id]
-    } else {
-      throw Error(`Package ${id} does not exist or the metadata is invalid`)
+    }
+    if (!allPackages[id] || !allPackages[id].metadata) {
+      throw new Error(`Invalid package ID: ${id}`)
     }
 
-    const orderedIds = [] as string[]
-    const currentStackClone = currentStack.slice()
-
-    allPackages[id].metadata.dependencies.forEach((dependency) => {
-      const ids: string[] = resolveDeps(dependency, currentStackClone)
-      orderedIds.push(...ids)
+    visited.add(id)
+    const deps = allPackages[id].metadata.dependencies || []
+    deps.forEach((dep) => {
+      if (!node[dep]) {
+        node[dep] = {}
+        addDependencies(dep, node[dep])
+      }
     })
-    orderedIds.push(id)
-    return orderedIds
+    visited.delete(id)
   }
 
-  let orderedPackageIds = [] as string[]
-  chosenPackageIds.forEach((packageId) => {
-    let packageIds = orderedPackageIds.concat(resolveDeps(packageId, []))
-    orderedPackageIds = packageIds.filter(
-      (id, index) => packageIds.indexOf(id) == index
-    )
+  chosenPackageIds.forEach((id) => {
+    if (!tree[id]) {
+      tree[id] = {}
+      addDependencies(id, tree[id])
+    }
   })
-  return orderedPackageIds
+  return tree
+}
+
+export const walkDependencyTree = async (tree, preOrPost, action) => {
+  const visitNode = async (node) => {
+    await Promise.all(
+      Object.keys(node).map(async (key) => {
+        if (preOrPost === 'pre') await action(key)
+        await visitNode(node[key])
+        if (preOrPost === 'post') await action(key)
+      })
+    )
+  }
+  await visitNode(tree)
+}
+
+export const concurrentifyAction = (
+  action: (id: string) => Promise<void>,
+  maxConcurrentActions: number
+) => {
+  const idToPromiseMap: Map<string, Promise<void>> = new Map()
+  const activePromises: Promise<void>[] = []
+
+  const concurrentAction = async (
+    id: string,
+    action: (id: string) => Promise<void>
+  ) => {
+    // Wait until there's space to start a new action
+    while (activePromises.length >= maxConcurrentActions) {
+      await Promise.race(activePromises)
+    }
```

Comment:
It won't be exhaustive as it awaits on a promise to finish then checks again. Under normal cases this will on;y loop once. We don't want to wait for all to finish only for one to finish which implies there is now space for another action to be running.